### ISSUE TITLE
[BE] Implementing Search

### ISSUE THREAD

**Issue by demet47** [bounswe/bounswe2022group7#547]
In our group meeting after the second milestone, we distributed the remaining tasks and ı was assigned to implement search in backend. 

**Deadline:** 19/12/2022
**Reviewer:** @sabrimete

**Comment by demet47** [comment]
I started implementing a version with external api call but it requires pagination etc. and would create a long response time. So instead I switched to a version where we utilize MYSQL full text search using natural language processing. 
Link for the latter can be found [here](https://www.w3resource.com/mysql/mysql-full-text-search-functions.php).

**Comment by demet47** [comment]
I chose to use full text search with option natural language mode. It is a feature in MySQL (and maybe some others too but not all contains it).

I put FULL TEXT indexes to necessary tables.
When one searches for keyword, below will be searched with the given fields as lookup fields:
- **Art Item:** name, description, category, labels
- **Event:** title, description, category, labels
- **Discussion Post:** title, text_body
- **Registered User:** name, surname, username, country

**Comment by demet47** [comment]
An example sql for creating full text index and sending a select query in naturala language mode:


```
ALTER TABLE artshare.art_item_info
    ADD FULLTEXT(name, description, category, labels);
```

```
SELECT * FROM art_item_info artinfo
WHERE MATCH (name, description,category ,labels)
             AGAINST ('sailing, boat' IN NATURAL LANGUAGE MODE);
```

**Comment by demet47** [comment]
There has been some slight modifications:

- There are seperate search endpoints for each data type now. I was sequentially appending the searches in a list by seperate calls anyway. This way enables filtering search in object domains also if wanted in frontend.
- I added an automatic initialization mechanism for adding indexes. HOW I did it can be found via below links:
    - [Stack overflow link](https://stackoverflow.com/questions/20964269/1214-the-used-table-type-doesnt-support-fulltext-indexes)
    - [Baeldung link](https://www.baeldung.com/spring-boot-data-sql-and-schema-sql)

**Comment by demet47** [comment]
Auto initialization created some errors in CICD for tests. It is related to application properties. there are seperate application.properties files for test folder and src folder but test folder by default reads from src's while initially being runned unfortunately. I browsed some pages but I couldn't find a solution. Thus I just added a file with necessary sql's to be run named run.sql under resources folder in case we drop the db.
